dynamiq/storages/vector/pgvector/pgvector.py

```diff
@@ -0,0 +1,648 @@
+    PGVectorVectorFunction.COSINE_SIMILARITY: "1 - ({embedding_key} <=> {query_embedding})",
+    PGVectorVectorFunction.INNER_PRODUCT: "({embedding_key} <#> {query_embedding}) * -1",
+    PGVectorVectorFunction.L2_DISTANCE: "{embedding_key} <-> {query_embedding}",
+    PGVectorVectorFunction.L1_DISTANCE: "{embedding_key} <+> {query_embedding}",
+}
+
+DEFAULT_TABLE_NAME = "dynamiq_vector_store"
+DEFAULT_SCHEMA_NAME = "public"
+
+
+class PGVectorStoreParams(BaseWriterVectorStoreParams):
+    embedding_key: str = "embedding"
+
+
+class PGVectorStore:
+    """Vector store using pgvector."""
+
+    def __init__(
+        self,
+        connection: PGVector | str | None = None,
+        client: Optional["PGVector"] = None,
+        create_extension: bool = True,
+        table_name: str | None = None,
+        schema_name: str | None = None,
+        dimension: int = 1536,
+        vector_function: PGVectorVectorFunction = PGVectorVectorFunction.COSINE_SIMILARITY,
+        index_method: PGVectorIndexMethod = PGVectorIndexMethod.EXACT,
+        index_name: str | None = None,
+        create_if_not_exist: bool = True,
+        content_key: str = "content",
+        embedding_key: str = "embedding",
+    ):
+        """
+        Initialize a PGVectorStore instance.
+
+        Args:
+            connection (PGVector | str): PGVector connection instance. Defaults to None.
+            client (Optional[PGVector]): PGVector client instance. Defaults to None.
+            create_extension (bool): Whether to create the vector extension (if it does not exist). Defaults to True.
+            table_name (str): Name of the table in the database. Defaults to None.
+            schema_name (str): Name of the schema in the database. Defaults to None.
+            dimension (int): Dimension of the embeddings. Defaults to 1536.
+            vector_function (PGVectorVectorFunction): The vector function to use for similarity calculations.
+                Defaults to 'cosine_similarity'.
+            index_method (PGVectorIndexMethod): The index method to use for the vector store.
+                Defaults to 'exact_nearest_neighbor_search'.
+            index_name (str): Name of the index to create. Defaults to None.
+            create_if_not_exist (bool): Whether to create the table and index if they do not exist. Defaults to True.
+            content_key (Optional[str]): The field used to store content in the storage. Defaults to 'content'.
+            embedding_key (Optional[str]): The field used to store embeddings in the storage. Defaults to 'embedding'.
+        """
+        if vector_function not in PGVectorVectorFunction:
+            raise ValueError(f"vector_function must be one of {PGVectorVectorFunction.list()}")
+        if index_method is not None and index_method not in PGVectorIndexMethod:
+            raise ValueError(f"index_method must be one of {PGVectorIndexMethod.list()}")
+
+        if isinstance(connection, str):
```

Comment:
It seems we should connect this just in case `if client is None`. client in you case is already created connection.

Reply:
Added the improved handling of connections for `PGVectorStore` aligned with your logic. Thanks, @acoola!